tensorflow_probability/python/experimental/mcmc/covariance_reducer.py

```diff
@@ -0,0 +1,260 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""CovarianceReducer for pre-packaged streaming covariance."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# Dependency imports
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.experimental.mcmc import reducer as reducer_base
+from tensorflow_probability.python.experimental.stats import sample_stats
+from tensorflow_probability.python.internal import prefer_static as ps
+from tensorflow_probability.python.mcmc.internal import util as mcmc_util
+from tensorflow.python.util import nest
+
+
+__all__ = [
+    'CovarianceReducer',
+    'VarianceReducer'
+]
+
+
+class CovarianceReducer(reducer_base.Reducer):
+  """`Reducer` that computes a running covariance.
+
+  As with all reducers, CovarianceReducer does not hold state information;
+  rather, it stores supplied metadata. Intermediate calculations are held in
+  a state object, which is returned via `initialize` and `one_step` method
+  calls.
+
+  `CovarianceReducer` is meant to fit into the larger Streaming MCMC
+  framework. `RunningCovariance` in `tfp.experimental.stats` is better suited
+  for more generic streaming covariance needs.
+  """
+
+  def __init__(self, event_ndims=None, ddof=0, name=None):
+    """Instantiates this object.
+
+    The running covariance computation supports batching. The `event_ndims`
+    parameter indicates the number of trailing dimensions to treat as part of
+    the event, and to compute covariance across. The leading dimensions, if
+    any, are treated as batch shape, and no cross terms are computed.
+
+    For example, if the incoming samples have shape `[5, 7]`, the `event_ndims`
+    selects among three different covariance computations:
+    - `event_ndims=0` treats the samples as a `[5, 7]` batch of scalar random
+      variables, and computes their variances in batch.  The shape of the result
+      is `[5, 7]`.
+    - `event_ndims=1` treats the samples as a `[5]` batch of vector random
+      variables of shape `[7]`, and computes their covariances in batch.  The
+      shape of the result is `[5, 7, 7]`.
+    - `event_ndims=2` treats the samples as a single random variable of
+      shape `[5, 7]` and computes its covariance.  The shape of the result
+      is `[5, 7, 5, 7]`.
+
+    For nested latent samples, both `event_ndims` and `ddof` must be either
+    a single value or an identical structure. For example, if the chain state
+    is the tuple of dictionaries ({pairA, pairB}, {pairC}), one could use
+    scalar values of `event_ndims` (integer or `None`) and `ddof` (integer) that
+    apply to all states in the latent. If not, `event_ndims` and `ddof` must
+    have structures that ressemble ({keyA: argA, keyB: argB}, {keyC: argC}).
```

Comment:
"resemble"

Reply:
Or rephase this, as it's not clear what "resemble" would mean here. `tf.nest` has very limited duck-typing of structures.